src/kokkos_types.hpp

```diff
@@ -0,0 +1,252 @@
+using ParArray4D = ParArrayGeneric<ParArray4DRaw<T>, State>;
+template <typename T, typename State = empty_state_t>
+using ParArray5D = ParArrayGeneric<ParArray5DRaw<T>, State>;
+template <typename T, typename State = empty_state_t>
+using ParArray6D = ParArrayGeneric<ParArray6DRaw<T>, State>;
+template <typename T, typename State = empty_state_t>
+using ParArray7D = ParArrayGeneric<ParArray7DRaw<T>, State>;
+template <typename T, typename State = empty_state_t>
+using ParArray8D = ParArrayGeneric<ParArray8DRaw<T>, State>;
+
+// Host mirrors
+template <typename T>
+using HostArray0D = typename ParArray0D<T>::HostMirror;
+template <typename T>
+using HostArray1D = typename ParArray1D<T>::HostMirror;
+template <typename T>
+using HostArray2D = typename ParArray2D<T>::HostMirror;
+template <typename T>
+using HostArray3D = typename ParArray3D<T>::HostMirror;
+template <typename T>
+using HostArray4D = typename ParArray4D<T>::HostMirror;
+template <typename T>
+using HostArray5D = typename ParArray5D<T>::HostMirror;
+template <typename T>
+using HostArray6D = typename ParArray6D<T>::HostMirror;
+template <typename T>
+using HostArray7D = typename ParArray7D<T>::HostMirror;
+
+using team_policy = Kokkos::TeamPolicy<>;
+using team_mbr_t = Kokkos::TeamPolicy<>::member_type;
+
+template <typename T>
+using ScratchPad1D = Kokkos::View<T *, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+template <typename T>
+using ScratchPad2D = Kokkos::View<T **, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+template <typename T>
+using ScratchPad3D = Kokkos::View<T ***, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+template <typename T>
+using ScratchPad4D = Kokkos::View<T ****, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+template <typename T>
+using ScratchPad5D = Kokkos::View<T *****, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+template <typename T>
+using ScratchPad6D = Kokkos::View<T ******, LayoutWrapper, ScratchMemSpace, MemUnmanaged>;
+
+// Used for ParArrayND
+// TODO(JMM): Should all of parthenon_arrays.hpp
+// be moved here? Or should all of the above stuff be moved to
+// parthenon_arrays.hpp?
+inline constexpr std::size_t MAX_VARIABLE_DIMENSION = 7;
+template <typename T, typename Layout = LayoutWrapper>
+using device_view_t =
+    Kokkos::View<multi_pointer_t<T, MAX_VARIABLE_DIMENSION>, Layout, DevMemSpace>;
+template <typename T, typename Layout = LayoutWrapper>
+using host_view_t = typename device_view_t<T, Layout>::HostMirror;
+
+template <typename ND, typename State = empty_state_t>
+struct ParArrayND_impl {
+  static_assert(ND::value <= 8, "ParArray only supported up to ND=8");
+};
+
+template <typename ND>
+struct ScratchPadND_impl {
+  static_assert(ND::value <= 6, "ScratchPad only supported up to ND=6");
+};
+
+template <std::size_t ND, typename T, typename State = empty_state_t>
+using ParArray = typename ParArrayND_impl<std::integral_constant<std::size_t, ND>,
+                                          State>::template type<T>;
+
```

Comment:
and below: this is a new interface, isn't it? Might be worth to briefly add this to the doc along the ParArray#D.

Reply:
yes, I've added some new documentation for it